### PR TITLE
Add flag for syntax-only config file checking in promtool

### LINKED ISSUE #5222: Make promtool better for CI with syntax-only config check
## Proposal
**Use case. Why is this important?**

*We validate our rules with promtool in CI before deploying, it would be nice to be able to check for config file syntax validity without having to deploy all secrets, certs, etc.*

## Bug Report
**What did you do?**
```promtool check config prometheus.yaml```

**What did you expect to see?**
```
./promtool check config --syntax-only prometheus.yaml                                                                                                                                                                                
Checking prometheus.yaml
  SUCCESS: prometheus.yaml is valid prometheus config file syntax
  SUCCESS: 0 rule files found
```
**What did you see instead? Under which circumstances?**
```
promtool check config prometheus.yaml
Checking prometheus.yaml
  FAILED: error checking bearer token file "/var/run/secrets/kubernetes.io/serviceaccount/token": stat /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
```

**Environment**
CI (explicitly does not have the token and certs present in the deployment)

* System information:

	Linux 3.10.0-514.26.2.el7.x86_64 x86_64

* Prometheus version:

	2.7.1

### PR DESCRIPTION
**Pull request by glightfoot**
Fixes #5222 by adding an optional --syntax-only flag to the check config command in promtool. Using this flag skips everything after checking to see if the file can be parsed by config.LoadFile(filename)

@brian-brazil

### CODE REVIEW

**Comment by simonpasquier** [comment]
You would need to merge/rebase on `master` to pass the CI.

**Comment by brian-brazil** [comment]
Is this something we want to add to every program we have that has a config file?

**Comment by glightfoot** [comment]
@brian-brazil - the only use case I can think of for it is promtool for checking assets in CI.

**Comment by brian-brazil** [comment]
I'd personally expect such a test to cover all the files referenced existing and being valid, so that it'll actually work when deployed.

**Comment by glightfoot** [comment]
This allows for both use cases - the default usage checks everything, while this syntax-only flag allows for CI to catch syntax errors without having to render out deployment-specific files every time. 

Default behavior - still checks all files referenced in the config
```
./promtool check config prometheus.yaml
Checking prometheus.yaml
  FAILED: error checking bearer token file "/var/run/secrets/kubernetes.io/serviceaccount/token": stat /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
```

New, optional behavior - just checks for valid syntax
```
./promtool check config --syntax-only prometheus.yaml                                                                                                                                                                                
Checking prometheus.yaml
  SUCCESS: prometheus.yaml is valid prometheus config file syntax
  SUCCESS: 0 rule files found
```

**Comment by glightfoot** [comment]
@simonpasquier - I merged master, but travis ci is still failing with the following error, any ideas?
```go: github.com/hashicorp/consul@v0.0.0-20180615161029-bed22a81e9fd: git fetch -f origin refs/heads/*:refs/heads/* refs/tags/*:refs/tags/* in /home/travis/gopath/pkg/mod/cache/vcs/45db37aaff0ad677e2105fd5f9d488e43f934feb6a5f8514e7fadf3de2b26381: exit status 128:
	error: RPC failed; curl 18 transfer closed with outstanding read data remaining
	fatal: The remote end hung up unexpectedly
...
go: error loading module requirements
make: *** [common-unused] Error 1
The command "make check_license style unused test staticcheck check_assets" exited with 2.
```

FWIW, make unused fails locally with the same error

**Comment by glightfoot** [comment]
Is there anything we can do to move this forward? This is useful to have as a linter

**Comment by glightfoot** [comment]
@krasi-georgiev @brian-brazil - is there anything we can do to address your concerns with this?

File: cmd/promtool/main.go [hunk 1]
```diff
@@ -52,6 +52,7 @@ func main() {
 		"config-files",
 		"The config files to check.",
 	).Required().ExistingFiles()
+	checkConfigSyntaxOnly := checkConfigCmd.Flag("syntax-only", "Only check the config file syntax; ignore certificate and token checks").Bool()
```
Comment (brian-brazil) [inline cmd/promtool/main.go]: I wouldn't mention certs or tokens (which come from a library in any case). It's files and their contents.

File: cmd/promtool/main.go [hunk 2]
```diff
@@ -194,54 +196,56 @@ func checkConfig(filename string) ([]string, error) {
...
-		if err := checkFileExists(scfg.HTTPClientConfig.BearerTokenFile); err != nil {
-			return nil, errors.Wrapf(err, "error checking bearer token file %q", scfg.HTTPClientConfig.BearerTokenFile)
+			ruleFiles = append(ruleFiles, rfs...)
 		}
 
-		if err := checkTLSConfig(scfg.HTTPClientConfig.TLSConfig); err != nil {
-			return nil, err
-		}
+		for _, scfg := range cfg.ScrapeConfigs {
+			if err := checkFileExists(scfg.HTTPClientConfig.BearerTokenFile); err != nil {
```
Comment (brian-brazil) [inline cmd/promtool/main.go]: This isn't a full list of the files.

Comment (glightfoot) [reply cmd/promtool/main.go]: I have not modified this section of code, other than wrapping it inside the `if !syntaxOnly {...}` block

File: cmd/promtool/main.go (hunk 1 above)
Comment (glightfoot) [reply cmd/promtool/main.go]: Thanks, I'll change the wording.

File: cmd/promtool/main.go (hunk 2 above)
Comment (brian-brazil) [reply cmd/promtool/main.go]: It's still something that needs fixing, and is related.

Comment (glightfoot) [reply cmd/promtool/main.go]: What's it missing?

Comment (brian-brazil) [reply cmd/promtool/main.go]: I make it consul openstack marathon triton alertmanager remote_write remote_read

The current approach clearly isn't very maintainable. Something smarter would be good.

**Comment by m0j0hn** [comment]
Any progress on merging this?

**Comment by toxick** [comment]
Running into the same need.  Would be nice to have this feature.

**Comment by glightfoot** [comment]
Ran into another failed deploy that this could prevent today. What can I do to help move this along? Is there a fundamental problem with the way syntax-only checking is implemented as is?

**Comment by brian-brazil** [comment]
There are still outstanding comments to address.

**Comment by glightfoot** [comment]
Closing in favor of #6488 . If we want to check more files, then I can open another issue and address it in another PR

**Pull request closed without merging**